### PR TITLE
OpenGL renderer a.k.a. shader support

### PR DESCRIPTION
**Pull request by ts-korhonen**
Summary
=======
I've been working on shader support via renderer that creates OpenGL context window allowing use of OpenGL library. It is now in a state where it seems to be working while there's still more to be done. Currently it replaces the existing SDL OpenGL option in menus and has no additional UI. I have selected the libretro shader files as a target for compatibility.

Implemented features in current state:
- Rendering in OpenGL (3.3 core)
- Basic shader support for libretro format shader files
  - Single-pass, no previous frames or scaling
  - path is hardcoded (shaders/shader.glsl)

Planned features:
- Menu options & saving them in config
- Reading .glslp files for settings
- Multiple shader passes
- Keeping previous frames

I didn't want to start modifying the UI yet and I opened this draft PR hoping to get some feedback from core developers.
Should this current state be merged into main repo, maybe in the experimental build?
Should it be it's own option and not replace the current SDL OpenGL option like it's doing now?

Checklist
=========
* [ ] Closes #xxx
* [ ] I have discussed this with core contributors already
* [ ] This pull request requires changes to the ROM set
  * [ ] I have opened a roms pull request - https://github.com/86Box/roms/pull/changeme/

References
==========
_Provide links to datasheets or other documentation that helped you implement this pull request._

### CODE REVIEW

**Comment by OBattler** [comment]
I do believe it should very much be merged into the repo, perhaps behind #ifdef DEV_BRANCH so only the Dev build has the feature, and I do agree with the idea of having it as a separate option.

**Comment by dhrdlicka** [comment]
I am concerned that having both SDL/OpenGL and pure OpenGL as available options might generate unnecessary confusion, especially if your OpenGL renderer supports shaders, which the existing SDL option does not. Is there any real benefit in keeping _both_?

File: src/win/win_opengl.c [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+static union
+{
+	struct
+	{
+		HANDLE closing;
+		HANDLE resize;
+		HANDLE blit_waiting;
+	};
+	HANDLE asArray[3];
+} sync_objects = {};
```
Comment (dhrdlicka) [inline src/win/win_opengl.c:93]: Zero initializing with `{}` is non-standard, please use `{0}`.

File: src/win/win_opengl.c [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+*/
+static HANDLE blit_done = NULL;
+
+/**
+ * @brief Blit event parameters.
+*/
+static volatile struct
+{
+	int x, y, y1, y2, w, h, resized;
+} blit_info = {};
```
Comment (dhrdlicka) [inline src/win/win_opengl.c:106]: Zero initializing with `{}` is non-standard, please use `{0}`.

File: src/win/win_opengl.c [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+
+/**
+ * @brief Resize event parameters.
+*/
+static volatile struct
+{
+	int width, height, fullscreen, scaling_mode;
+} resize_info = {};
```
Comment (dhrdlicka) [inline src/win/win_opengl.c:114]: Zero initializing with `{}` is non-standard, please use `{0}`.

**Comment by OBattler** [comment]
The current OpenGL (which is 2.0, I believe) in some case performs better than 3.0, so I do think there should be a choice.

File: src/win/win_opengl_glslp.c [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+#include <glad/glad.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <86box/86box.h>
+#include <86box/plat.h>
```
Comment (dhrdlicka) [inline src/win/win_opengl_glslp.c:39]: It appears that these includes fail in a pretty weird way for MSVC builds:
```
  win_opengl_glslp.c
D:\a\86Box\86Box\src\include\86box/86box.h(72,17): error C2061: syntax error: identifier 'unique_id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(72,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(73,17): error C2061: syntax error: identifier 'source_hwnd' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(73,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(104,17): error C2061: syntax error: identifier 'mem_size' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(104,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(176,17): error C2061: syntax error: identifier 'get_last_addr' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(176,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/86box.h(176,35): error C2059: syntax error: '<parameter-list>' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(74,17): error C2061: syntax error: identifier 'timer_freq' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(74,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(102,17): error C2061: syntax error: identifier 'plat_timer_read' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(102,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(102,37): error C2059: syntax error: '<parameter-list>' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(103,17): error C2061: syntax error: identifier 'plat_get_ticks' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(103,17): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(103,36): error C2059: syntax error: '<parameter-list>' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(104,36): error C2146: syntax error: missing ')' before identifier 'count' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(104,36): error C2061: syntax error: identifier 'count' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(104,36): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(104,41): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(131,34): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(131,34): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(131,34): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(131,36): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(131,58): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(132,34): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(132,34): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(132,34): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(132,36): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(133,33): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(133,33): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(133,33): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(133,35): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(133,45): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(134,42): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(134,42): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(134,42): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(134,44): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(134,60): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(135,31): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(135,31): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(135,31): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(135,33): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(136,31): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(136,31): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(136,31): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(136,33): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(136,55): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(137,32): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(137,32): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(137,32): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(137,34): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(138,30): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(138,30): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(138,30): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(138,32): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(139,30): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(139,30): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(139,30): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(139,32): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(139,54): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(140,31): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(140,31): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(140,31): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(140,33): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(141,36): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(141,36): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(141,36): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(141,38): error C2059: syntax error: ',' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(141,46): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(142,37): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(142,37): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(142,37): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(142,39): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(143,37): error C2146: syntax error: missing ')' before identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(143,37): error C2061: syntax error: identifier 'id' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(143,37): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(143,39): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(174,34): error C2146: syntax error: missing ')' before identifier 'dst' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(174,34): error C2061: syntax error: identifier 'dst' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(174,34): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(174,37): error C2059: syntax error: '[' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(174,66): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(175,52): error C2146: syntax error: missing ')' before identifier 'src' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(175,52): error C2061: syntax error: identifier 'src' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(175,52): error C2059: syntax error: ';' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(175,55): error C2059: syntax error: '[' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
D:\a\86Box\86Box\src\include\86box/plat.h(175,66): error C2059: syntax error: ')' [D:\a\86Box\86Box\build\src\win\ui.vcxproj]
```

**Review by dhrdlicka — COMMENTED** [review]
I noticed that the MSVC build is broken -- if you could fix it then it would be awesome, though it's not really that important.

File: src/win/win_opengl_glslp.c (hunk 4 above)
Comment (ts-korhonen) [reply src/win/win_opengl_glslp.c:39]: yes, I'm just checking this out and can't seem to notice what's wrong. The other ones was just about C flavor too new for MSVC.

**Comment by dhrdlicka** [comment]
> The current OpenGL (which is 2.0, I believe) in some case performs better than 3.0, so I do think there should be a choice.

In that case it might be better to adapt the new renderer from this PR so there is feature parity.

@ts-korhonen would it be possible/easy to add OpenGL 2.0 support to your current code?

**Comment by ts-korhonen** [comment]
> @ts-korhonen would it be possible/easy to add OpenGL 2.0 support to your current code?
I don't think so, it's OpenGL Core that is newer. I think keeping the old one for that is better.

File: src/win/win_opengl_glslp.c (hunk 4 above)
Comment (ts-korhonen) [reply src/win/win_opengl_glslp.c:39]: Ok, needed #include <stdint.h>
It should probably be added to 86box.h and plat.h too

**Comment by OBattler** [comment]
> I don't think so, it's OpenGL Core that is newer. I think keeping the old one for that is better.
I agree. The old one should be kept in addition to the new one.

**Pull request merged**